File: packages/react-native/src/protocol/snapshot.ts
/**
 * The mobile consent snapshot: the one immutable value the native cores own
 * and the JavaScript layer renders.
 *
 * This is the shape on the wire. It is the kernel snapshot from `@c15t/core`
 * restricted to the fields that matter on device, plus the lifecycle flags a
 * native gate needs (`ready`, `policyPending`) and the IAB slot, which carries
 * the vendor list a device was served. The native cores serialize exactly this
 * object from `getSnapshot()`; the JavaScript layer never derives it.
 */

import type { GlobalVendorList } from './gvl';
import type { NativeOverrides } from './overrides';
import type {
	AllConsentNames,
	ConsentState,
	ConsentSubject,
	ExplicitChoice,
	KernelActiveUI,
	KernelModel,
	KernelTranslations,
	LocationResponse,
	OptionalConsentCategory,
	PolicyResolutionStatus,
	PromptRequirement,
	RestrictionReason,
} from './vocabulary';

/**
 * Consent model enforced on device.
 *
 * `iab` is excluded because no native core resolves it: Kotlin's `Enums.kt`
 * refuses the wire value and `StrictPolicyReader` fails closed on it, and
 * Swift's `ConsentModel` never names it. A core that answered `iab` would
 * promise an evaluation it cannot do.
 *
 * Reading a served vendor list is a different question, and {@link
 * NativeIABState} answers it: a device may hold the list `/init` embedded while
 * the model it evaluates is `opt-in`, because the list is disclosure, not a
 * permission rule.
 */
export type NativeModel = Exclude<KernelModel, 'iab'>;

/**
 * Explainability half of the policy resolution, as carried on the snapshot.
 *
 * The matched rule itself stays on the native side. The JavaScript layer only
 * needs to know whether a rule matched and which one, so it can show
 * explainability and decide when a re-prompt is owed.
 */
export interface SnapshotResolution {
	/** Resolution outcome, shared with the kernel's vocabulary. */
	readonly status: PolicyResolutionStatus;
	/** Identifier of the matched policy, `null` when nothing matched. */
	readonly policyId: string | null;
	/** Fingerprint of the matched rule, `null` when nothing matched. */
	readonly fingerprint: string | null;
}

/**
 * Privacy signals the native cores honor.
 *
 * Mirrors `KernelPrivacySignals` in `@c15t/core`. The three members are not
 * interchangeable and a gate must read `active`:
 *
 * - `detected` is what the device reported. On mobile there is no user-agent
 *   flag to read, so this is the host app's report, or what `/init` resolved.
 * - `override` is {@link NativeOverrides.gpc}, `null` when the app set none.
 * - `active` is the value the evaluator honors: the override when set, else
 *   the detection.
 */
export interface NativeGpcSignal {
	/** Signal the device or backend reported. */
	readonly detected: boolean;
	/** Explicit override from {@link NativeOverrides.gpc}, or `null`. */
	readonly override: boolean | null;
	/** Signal the evaluator honors: the override when set, else the detection. */
	readonly active: boolean;
}

/**
 * Privacy signals the native cores honor.
 */
export interface NativePrivacySignals {
	readonly gpc: NativeGpcSignal;
}

/**
 * Failure the native core is holding, if any.
 *
 * A failed init does not throw into the app. It sets `error`, keeps
 * `policyPending` true, and leaves every optional category denied.
 */
export interface NativeSnapshotError {
	readonly code: string;
	readonly message: string;
}

/**
 * The IAB half of a snapshot.
 *
 * Mirrors `KernelIABState` in `@c15t/core` on `gvl` alone. The kernel's other
 * members -- `authority`, `enabled`, `cmpId`, `tcString`, `gvlReference`,
 * `customVendors`, and the per-vendor and per-purpose vectors -- are IAB runtime
 * state a device does not own yet, and an absent key says that plainly where
 * `enabled: false` would claim an answer about a module that is not there.
 * Swift's `KernelIABState` makes the same cut for the same reason and encodes
 * `gvl` always present, so a reader never confuses "no list served" with "no
 * answer given". The document behind the name is `./gvl`.
 */
export interface NativeIABState {
	/** Global Vendor List (IAB-registered vendors + purposes), or `null`. */
	readonly gvl: GlobalVendorList | null;
}

/**
 * The immutable snapshot `getSnapshot()` returns as a JSON string.
 *
 * `ready` and `policyPending` are the two flags a gate must consult: while
 * either is unset, every optional category reads `false`. `revision` is
 * monotonic and is the only value an event needs to carry, because
 * subscribers pull the snapshot instead of receiving a copy of it.
 */
export interface ConsentSnapshot {
	/** Monotonic revision, bumped on every native mutation. */
	readonly revision: number;
	/** `true` until the first init resolves a policy. */
	readonly policyPending: boolean;
	/** `false` until the stored envelope finished hydrating. */
	readonly ready: boolean;
	/** Model the evaluator applies. */
	readonly model: NativeModel;
	/** Surface the UI should render, if any. */
	readonly activeUI: KernelActiveUI;
	/** Interaction the active policy still requires. */
	readonly promptRequirement: PromptRequirement;
	/** Effective permissions every gate reads. */
	readonly effectivePermissions: ConsentState;
	/** Latest per-category receipts. Only accept, reject and save write it. */
	readonly explicitChoice: ExplicitChoice | null;
	/**
	 * Categories the consent surfaces list: `necessary` first, then the categories
	 * of the resolved policy scope the app's declared scope survives. The native
	 * core decides this list the way the web dialog decides its own, so a name the
	 * policy does not govern is never rendered. `null` only while no configuration
	 * has been installed, where the UI lists every category it knows.
	 */
	readonly consentCategories: readonly AllConsentNames[] | null;
	/** Categories denied regardless of grants, with the reason. */
	readonly restrictions: Partial<
		Record<OptionalConsentCategory, readonly RestrictionReason[]>
	>;
	/** Explainability for how the policy was matched. */
	readonly resolution: SnapshotResolution;
	/** Signed token for write-time consistency, sent back on save. */
	readonly policySnapshotToken: string | null;
	/** Subject identifiers, carried once. */
	readonly subject: ConsentSubject | null;
	/** Geographic context reported by the backend. */
	readonly location: LocationResponse | null;
	/** Geographic, language, and GPC overrides in effect. */
	readonly overrides: NativeOverrides;
	/** Effective privacy signals. */
	readonly privacySignals: NativePrivacySignals;
	/** Standing privacy directives. Always empty in this phase. */
	readonly optOutDirectives: readonly [];
	/** Resolved translation bundle. */
	readonly translations: KernelTranslations | null;
	/** Earliest future time (epoch ms) that can change permissions. */
	readonly nextDeadline: number | null;
	/** Epoch milliseconds of the last evaluation. */
	readonly evaluatedAt: number;
	/** Failure the core is holding, or `null`. */
	readonly error: NativeSnapshotError | null;
	/**
	 * IAB TCF state: the vendor list this device was served, or `null`.
	 *
	 * `null` is a real answer, and there is one way to hold it: no `/init` has served a
	 * `gvl` this build could read. That is the ordinary answer for every policy whose
	 * model is not `iab`, because the backend embeds a list only while it is.
	 *
	 * Both cores fold a served list onto this key and keep it when a later `/init`
	 * serves none. The disclosure a subject was already shown is drawn out of that
	 * document, so dropping it would change a claim the subject was given without the
	 * subject doing anything, and only a reset takes it back. Kotlin keeps the bytes
	 * under the stored envelope's own `gvl` key and puts them back here on hydration,
	 * so the encrypted blob carries one copy of the largest thing in it and a device
	 * upgrading from a build that kept the list only there keeps its disclosure.
	 *
	 * A host app's own code reads the same document off the same state: Kotlin's
	 * `C15tKernel.vendorListBody()` and Swift's `globalVendorList()` are both the
	 * snapshot field, which is what keeps a disclosure drawn natively and one drawn
	 * through this bridge naming the same vendors.
	 *
	 * The key is never absent: a serialised `null` where there is no list is what lets
	 * a JavaScript layer that predates TCF avoid branching on an SDK version.
	 */
	readonly iab: NativeIABState | null;
}

File: native/core-android/c15t-core/src/main/kotlin/com/c15t/core/model/Snapshot.kt
package com.c15t.core.model

import com.c15t.core.tc.GlobalVendorList
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonObject

/**
 * Interaction the active policy still requires.
 *
 * `notice` is the first-layer prompt (banner), `acknowledge` the explicit
 * choice. `purpose` distinguishes a first prompt from a re-prompt caused by an
 * expired or changed policy.
 */
@Serializable
data class PromptRequirement(
	val notice: Boolean = false,
	val acknowledge: Boolean = false,
	val purpose: PromptPurpose? = null,
) {
	companion object {
		/** Nothing owed. */
		val NONE = PromptRequirement()
	}
}

/**
 * Policy resolution outcome carried on the snapshot.
 *
 * [status] is the raw wire status: `matched`, `no-match`, `unconfigured`, or
 * `failed`. [reason] is only present for `failed` and names why, mirroring
 * `PolicyResolutionFailure` in `@c15t/schema`.
 */
@Serializable
data class PolicyResolution(
	val status: String = STATUS_UNCONFIGURED,
	val policyId: String? = null,
	val fingerprint: String? = null,
	val reason: String? = null,
) {
	val matched: Boolean
		get() = status == STATUS_MATCHED

	companion object {
		const val STATUS_MATCHED = "matched"
		const val STATUS_NO_MATCH = "no-match"
		const val STATUS_UNCONFIGURED = "unconfigured"
		const val STATUS_FAILED = "failed"

		const val REASON_INVALID_CONFIGURATION = "invalid-configuration"
		const val REASON_INSUFFICIENT_INPUTS = "insufficient-inputs"
		const val REASON_TRANSPORT = "transport"
		const val REASON_UNSUPPORTED_CONTRACT = "unsupported-contract"
		const val REASON_INVALID_PAYLOAD = "invalid-payload"

		/** A `failed` resolution with [reason], carrying no policy. */
		fun failed(reason: String): PolicyResolution = PolicyResolution(
			status = STATUS_FAILED,
			reason = reason,
		)
	}
}

/**
 * Subject identifiers.
 *
 * [id] is the c15t-owned id minted at first launch by `SubjectIdGenerator`, in the
 * `sub_` format the backend's `subjectIdSchema` accepts. A stored id is adopted
 * untouched when the producer would take it and refused when it would not: an id every
 * save gets `INPUT_VALIDATION_FAILED` over is not an identity, and the envelope written
 * under it is worth the same nothing. See `SubjectIdGenerator.isValid` and
 * `native/CONTRACT.md`, Subject identity. Never derived from IDFV, ADID, or any other
 * hardware identifier.
 */
@Serializable
data class ConsentSubject(
	val id: String,
	val externalId: String? = null,
)

/** Geographic context reported by the backend. */
@Serializable
data class ConsentLocation(
	val country: String? = null,
	val region: String? = null,
	val language: String? = null,
)

/**
 * Developer overrides that steer policy evaluation.
 *
 * [gpc] is the app's override, not a detection. It is load bearing: `@c15t/core`
 * compares it against the decision inputs remembered from the last init and
 * rejects a save whose inputs no longer match, so a save body without it cannot
 * pass the backend's staleness check. Detection lives on [PrivacySignals.gpc].
 *
 * Publisher test mode is a client option rather than an override and never
 * reaches a save body, so it has no member here. The first draft of
 * `native/CONTRACT.md` invented a `test` field; [com.c15t.core.store.C15tStore]
 * refuses an envelope that still carries one rather than guessing at it.
 */
@Serializable
data class KernelOverrides(
	val country: String? = null,
	val region: String? = null,
	val language: String? = null,
	val gpc: Boolean? = null,
)

/**
 * The `gpc` member of [PrivacySignals]: the Global Privacy Control signal in the
 * three parts the kernel keeps apart.
 *
 * An evaluator reads [active]. [detected] and [override] exist so a host can tell
 * why a signal is on, and so a write is not treated as stale just because a
 * device started reporting GPC on its own.
 *
 * [active] is an output, never an input. A plain `copy` would let a caller set
 * `active: true` beside `detected: false` and no override, which is a signal the
 * two inputs do not support, so [withDetection] and [withOverride] are the only
 * ways to change one. The kernel recomputes it on every read rather than trusting
 * a stored copy, so an envelope that claims a signal is on while saying nothing
 * detected it and nothing overrode it is answered with the value the other two
 * support.
 */
@Serializable
data class GpcSignal(
	val detected: Boolean = false,
	val override: Boolean? = null,
	val active: Boolean = false,
) {
	/** Re-derive `active` from [override], keeping the reported detection. */
	fun withOverride(override: Boolean?): GpcSignal = derive(override = override, detected = detected)

	/** Re-derive `active` from [detected], keeping the app's override if any. */
	fun withDetection(detected: Boolean): GpcSignal = derive(override = override, detected = detected)

	companion object {
		/** Derive the view the way `derivePrivacySignals` in `@c15t/core` does. */
		fun derive(override: Boolean?, detected: Boolean): GpcSignal = GpcSignal(
			detected = detected,
			override = override,
			active = override ?: detected,
		)
	}
}

/**
 * Privacy signals the core honors, mirroring `KernelPrivacySignals`.
 *
 * There is no `msa` signal anywhere in v3. The first draft of
 * `native/CONTRACT.md` carried one beside a boolean `gpc`; both are refused when
 * a stored envelope is read.
 */
@Serializable
data class PrivacySignals(
	val gpc: GpcSignal = GpcSignal.derive(override = null, detected = false),
)

/**
 * User identification attached to consent records. External id only: no PII
 * beyond what the integrator opts in to.
 */
@Serializable
data class KernelUser(
	val externalId: String,
	val identityProvider: String? = null,
	val properties: Map<String, String>? = null,
)

/**
 * Standing privacy directive recorded from a device signal. It is a privacy
 * request, not a consent record, and it outlives the live signal.
 */
@Serializable
data class OptOutDirective(
	val source: String,
	val categories: List<String> = emptyList(),
	val recordedAt: Long,
)

/** A failure the core wants the host to surface. */
@Serializable
data class KernelError(
	val code: String,
	val message: String,
)

/**
 * Explicit category choices, captured once per commit.
 *
 * [consents] holds the explicit value per optional category; absent keys are
 * undecided and the evaluator fills them from the policy default. [fingerprint]
 * is the choice-prompt fingerprint the choice was made against, so a policy
 * change invalidates it instead of silently reusing an old grant.
 */
@Serializable
data class ExplicitChoice(
	val consents: Map<String, Boolean>,
	val action: ConsentAction,
	val actionAt: Long,
	val fingerprint: String? = null,
	val version: Int = 3,
) {
	/** Read the explicit value for [category], `null` when undecided. */
	fun valueOf(category: ConsentCategory): Boolean? = consents[category.wireName]

	/** `true` when the choice was made against [currentFingerprint]. */
	fun matchesFingerprint(currentFingerprint: String?): Boolean {
		if (fingerprint == null || currentFingerprint == null) {
			return true
		}
		return fingerprint == currentFingerprint
	}
}

/** The categories one commit confirmed, with the single captured action time. */
@Serializable
data class ConfirmedCoverage(
	val categories: Map<String, Boolean>,
	/** Epoch milliseconds captured once, before any disk or network call. */
	val actionAt: Long,
)

/** Resolved policy inputs captured with an action, replayed unchanged on retry. */
@Serializable
data class DecisionInputs(
	val policyId: String? = null,
	val fingerprint: String? = null,
	val country: String? = null,
	val region: String? = null,
	val language: String? = null,
	val gpc: Boolean = false,
)

/**
 * The IAB half of a [ConsentSnapshot]: the vendor list this device was served.
 *
 * The key name is `KernelIABState`'s, from `packages/core/src/types.ts`, because
 * a renamed key is invisible across the bridge: the snapshot crosses into
 * JavaScript as text, so the declared type can say `iab.gvl` while the device
 * sends `iab.list`, and every app on that platform reads `undefined` with nothing
 * failing anywhere. Only `gvl` is carried. `authority`, `tcString`, `cmpId` and
 * the per-vendor vectors are IAB runtime state this core does not own, and an
 * absent key says that plainly where `enabled: false` would claim an answer about
 * a module that is not there.
 *
 * `gvl` is non-null, which is stricter than the web's nullable key on purpose: this
 * core only ever builds the container around a list it accepted, so a non-null `iab`
 * always means a list is held and a reader never has to tell "no list served" apart
 * from "no answer given". The price is that stored bytes reading `{"gvl":null}` --
 * which `core-swift` takes as a container holding no list -- are unreadable here,
 * and neither core writes them. Absence is carried by [ConsentSnapshot.iab] being
 * `null`, which is also what `packages/react-native/src/protocol/wire-shape.ts`
 * grades.
 *
 * The spellings inside the list are the served document's, not this core's: see
 * [GlobalVendorList] and [com.c15t.core.tc.GlobalVendorListJson].
 */
@Serializable
data class KernelIabState(
	val gvl: GlobalVendorList,
)

/**
 * The one immutable consent state the native core holds, versioned by
 * [revision].
 *
 * Field set and semantics follow `native/CONTRACT.md`, which is the mobile
 * projection of `ConsentSnapshot` in `packages/core/src/types.ts`. The IAB half
 * is the served vendor list on [iab], the same seam `KernelIABState` is on the
 * web and in `core-swift`.
 * Every mutation produces a new instance and bumps [revision]; readers never see
 * a partially updated snapshot.
 */
@Serializable
data class ConsentSnapshot(
	/** Monotonic revision, bumps on every mutation. */
	val revision: Long = 0,
	/** `true` until the first init resolves a policy the core can represent. */
	val policyPending: Boolean = true,
	/** `false` until a stored snapshot was hydrated or an init resolved. */
	val ready: Boolean = false,
	val model: ConsentModel = ConsentModel.OPT_IN,
	val activeUI: ActiveUI? = ActiveUI.NONE,
	val promptRequirement: PromptRequirement = PromptRequirement.NONE,
	val effectivePermissions: ConsentState = ConsentState.DENY_ALL,
	val explicitChoice: ExplicitChoice? = null,
	/** Configured categories; `null` uses the full policy scope. */
	val consentCategories: List<String>? = null,
	/** Category name to restriction reasons, for explainability. */
	val restrictions: Map<String, List<String>> = emptyMap(),
	val resolution: PolicyResolution = PolicyResolution(),
	/** Signed token for write-time consistency, sent back on save. */
	val policySnapshotToken: String? = null,
	val subject: ConsentSubject? = null,
	val location: ConsentLocation? = null,
	val overrides: KernelOverrides = KernelOverrides(),
	val privacySignals: PrivacySignals = PrivacySignals(),
	val optOutDirectives: List<OptOutDirective> = emptyList(),
	/** Opaque translation bundle from `/init`, carried through untouched. */
	val translations: JsonObject? = null,
	/** Earliest future time (epoch ms) that can change permissions or the prompt. */
	val nextDeadline: Long? = null,
	/** Epoch milliseconds of the last evaluation. */
	val evaluatedAt: Long = 0,
	val error: KernelError? = null,
	/**
	 * The vendor list this device is reading its disclosure out of, or `null` while
	 * no `/init` has served one this build could accept.
	 *
	 * Latched, and it travels with the snapshot rather than beside it: a `/init`
	 * that serves no list leaves the list already there, and only
	 * [com.c15t.core.C15tKernel.reset] takes it back. The reason is on that field --
	 * once a purpose or vendor name has been on screen, replacing it with nothing
	 * changes a claim the subject was already given, and the backend embeds a list
	 * only while the matched model is `iab`.
	 *
	 * The key is always present on the wire, `null` written where there is no list,
	 * so a JavaScript layer that predates TCF does not branch on the SDK version.
	 * Storage keeps the document once, under the envelope's own `gvl` key, and puts
	 * it back here on hydration -- see
	 * [com.c15t.core.store.SnapshotEnvelope.gvl].
	 */
	@SerialName("iab")
	val iab: KernelIabState? = null,
) {
	/**
	 * Whether [category] may run right now.
	 *
	 * `necessary` is always granted. While [ready] or [policyPending] is unset
	 * every optional category reads `false`, which is the deny-all contract both
	 * flags guard.
	 */
	fun isAllowed(category: ConsentCategory): Boolean {
		if (!category.optional) {
			return true
		}
		if (!ready || policyPending) {
			return false
		}
		return effectivePermissions[category]
	}

	/**
	 * Whether this snapshot is in a state that answers for a decision: hydrated, and
	 * with the first policy resolution folded in.
	 *
	 * The two flags stay separate fields because the evaluator and the wire need them
	 * apart. This is the pair a host SDK reads as one question, and it is computed
	 * rather than stored so it cannot contradict the flags it comes from.
	 */
	val isReady: Boolean
		get() = ready && !policyPending

	/**
	 * Why [category] may or may not run, as the three states a native SDK gate
	 * reports.
	 *
	 * Derived from this snapshot and nothing else, so a gate can never disagree with
	 * the state the UI is showing. [ConsentDecision.PENDING] is what [isReady] false
	 * answers for every optional category: the device has not been told yet, so the
	 * `false` in [effectivePermissions] is not a refusal and must not end the wait.
	 */
	fun decision(category: ConsentCategory): ConsentDecision {
		if (!category.optional) {
			return ConsentDecision.GRANTED
		}
		if (!isReady) {
			return ConsentDecision.PENDING
		}
		return if (effectivePermissions[category]) {
			ConsentDecision.GRANTED
		} else {
			ConsentDecision.DENIED
		}
	}

	companion object {
		/** The snapshot a cold start with nothing stored answers with. */
		fun denyAll(subject: ConsentSubject? = null, now: Long = 0): ConsentSnapshot = ConsentSnapshot(
			policyPending = true,
			ready = false,
			effectivePermissions = ConsentState.DENY_ALL,
			subject = subject,
			evaluatedAt = now,
		)
	}
}
